virtcontainers/acrn_arch_base.go

```diff
@@ -0,0 +1,762 @@
+	// UUID is the acrn process UUID.
+	UUID string
+
+	// Devices is a list of devices for acrn to create and drive.
+	Devices []Device
+
+	// Kernel is the guest kernel configuration.
+	Kernel Kernel
+
+	// Memory is the guest memory configuration.
+	Memory Memory
+
+	acrnParams []string
+
+	// ACPI virtualization support
+	ACPIVirt bool
+
+	// NumCPU is the number of CPUs for guest
+	NumCPU uint32
+}
+
+// MaxAcrnVCPUs returns the maximum number of vCPUs supported
+func MaxAcrnVCPUs() uint32 {
+	return uint32(8)
+}
+
+func newAcrnArch(config HypervisorConfig) acrnArch {
+
+	a := &acrnArchBase{
+		path:                 acrnPath,
+		ctlpath:              acrnctlPath,
+		kernelParamsNonDebug: acrnKernelParamsNonDebug,
+		kernelParamsDebug:    acrnKernelParamsDebug,
+		kernelParams:         acrnKernelParams,
+	}
+
+	a.handleImagePath(config)
+	return a
+}
+
+func (a *acrnArchBase) acrnPath() (string, error) {
+	p := a.path
+	return p, nil
+}
+
+func (a *acrnArchBase) acrnctlPath() (string, error) {
+	ctlpath := a.ctlpath
+	return ctlpath, nil
+}
+
+func (a *acrnArchBase) kernelParameters(debug bool) []Param {
+	params := a.kernelParams
+
+	if debug {
+		params = append(params, a.kernelParamsDebug...)
+	} else {
+		params = append(params, a.kernelParamsNonDebug...)
+	}
+
+	return params
+}
+
+func (a *acrnArchBase) memoryTopology(memoryMb uint64) Memory {
+
+	mem := fmt.Sprintf("%dM", memoryMb)
+	memory := Memory{
+		Size: mem,
+	}
+
+	return memory
+}
+
+func (a *acrnArchBase) capabilities() types.Capabilities {
+	var caps types.Capabilities
+
+	// For devicemapper disable support for filesystem sharing
+
+	caps.SetFsSharingUnsupported()
+	caps.SetBlockDeviceSupport()
+	caps.SetBlockDeviceHotplugSupport()
+
+	return caps
+}
+
+// Valid returns true if the CharDevice structure is valid and complete.
+func (cdev ConsoleDevice) Valid() bool {
+
+	if cdev.Backend != "tty" && cdev.Backend != "pty" &&
+		cdev.Backend != "console" && cdev.Backend != "socket" &&
+		cdev.Backend != "file" {
+		return false
+	}
+
+	return true
+}
+
+// AcrnParams returns the acrn parameters built out of this console device.
+func (cdev ConsoleDevice) AcrnParams(slot int, config *Config) []string {
+
+	var acrnParams []string
+	var deviceParams []string
+
+	acrnParams = append(acrnParams, "-s")
+	deviceParams = append(deviceParams, fmt.Sprintf("%d,virtio-console,", slot))
+
+	if cdev.PortType == ConsoleBE {
+		deviceParams = append(deviceParams, "@")
+	}
+
+	switch cdev.Backend {
+	case "pty":
+		deviceParams = append(deviceParams, "pty:pty_port")
+	case "tty":
+		deviceParams = append(deviceParams, fmt.Sprintf("tty:tty_port=%s", cdev.Path))
+	case "socket":
+		deviceParams = append(deviceParams, fmt.Sprintf("socket:%s=%s", cdev.Name, cdev.Path))
+	case "file":
+		deviceParams = append(deviceParams, fmt.Sprintf("file:file_port=%s", cdev.Path))
+	case "stdio":
+		deviceParams = append(deviceParams, "stdio:stdio_port")
+	default:
+		// do nothing. Error should be already caught
```

Comment:
Maybe this case should return an error for "an impossible situation"? The function doesn't currently return one but maybe it should generally. For example, what if `cdev.Path == ""`? We've seen bugs like this in the past.

Reply:
the switch case is based on cdev.Backend and this is already validated in the corresponding Valid(). This function just appends the acrn devices to the command line and even if error is returned no action can be taken. Valid() function for each device will take care of error checking.
I have added cdev.path check in Valid() function.